### PR TITLE
[add] support for pages

### PR DESCRIPTION
**Pull request by Achoobert**
https://github.com/digi-serve/plugin_ABDesigner/issues/3

### CODE REVIEW

**Review by Hiro-Nakamura — CHANGES_REQUESTED** [review]
Yeah, make sure you have the latest copy in your `ab_platform_web`.

File: views/ABViewCore.js [hunk 1]
```diff
@@ -798,7 +798,11 @@ module.exports = class ABViewCore extends ABMLClass {
       let result = this.viewNew(config, this.application, parent);
 
       // change id
-      result.id = lookUpIds[result.id] || this.AB.uuid();
+      if (parent == null) {
```
[the remark this answers is not in the document]
Comment (Achoobert) [reply views/ABViewCore.js:801]: This is accounting for a case when the page is getting cloned to root: 
there is no parent, as the parent page is the application.

File: views/ABViewCore.js [hunk 2]
```diff
@@ -868,15 +874,19 @@ module.exports = class ABViewCore extends ABMLClass {
       result.parent = parent || this.parent;
 
       // change id
-      result.id = lookUpIds[result.id] || this.AB.uuid();
+      if (parent == null) {
```
[the remark this answers is not in the document]
Comment (Achoobert) [reply views/ABViewCore.js:877]: yes, this definitely needs some comments to explain what is happening.
iirc the page-adder will use that ID to find the parent the User selected on my new popup box.
If I don't pass any id, the recursive page-inserter defaults to adding the new page the the root. 
---the page-inserter/updater function always ticks through all the pages until it hits the one on root. The root page always has a null 'parent.' 
Here I'm just taking advantage of that existing function to drop new copied pages on root by setting this null

File: views/ABViewPageCore.js [hunk 3]
```diff
@@ -176,6 +176,13 @@ module.exports = class ABViewPageCore extends ABViewContainer {
          })
          .then(() => {
             var parent = this.parent || this.application;
+
+            // Force application to be able to insert page.
+            // This may not be a good solution?
+            if (!this.parent){
+              this.application.pageRemove = this.pageRemove
```
[the remark this answers is not in the document]
Comment (Achoobert) [reply views/ABViewPageCore.js]: Checked, it works, and removed

File: views/ABViewPageCore.js [hunk 4]
```diff
@@ -219,6 +226,13 @@ module.exports = class ABViewPageCore extends ABViewContainer {
          .then(() => {
             // now we can persist ourself in our parent
             var parent = this.parent || this.application;
+
+            // Force application to be able to insert page.
+            // This may not be a good solution?
+            if (!this.parent){
+              this.application.pageInsert = this.pageInsert
```
[the remark this answers is not in the document]
Comment (Achoobert) [reply views/ABViewPageCore.js]: Checked, it works, and removed

File: ABDefinitionCore.js [hunk 5]
```diff
@@ -26,8 +26,14 @@ module.exports = class ABDefinitionCore {
...
       }
       this.name =
-         attributes.name || attributes.json.name || attributes.json.label || "";
-      this.type = attributes.type || attributes.json.type || "";
+         attributes?.name ||
+         attributes?.json?.name ||
+         attributes?.json?.label ||
+         "";
+      if (!this.name) {
+         console.error("no name");
```
[the remark this answers is not in the document]
Comment (Achoobert) [reply ABDefinitionCore.js]: Update

File: ABApplicationCore.js [hunk 6]
```diff
@@ -856,6 +856,17 @@ module.exports = class ABApplicationCore extends ABMLClass {
       return ABViewManager.newView(values, this, null);
    }
 
+    /**
+    * @method pageByID()
+    * return the object's field from the given {ABField.id}
```
[the remark this answers is not in the document]
Comment (Achoobert) [reply ABApplicationCore.js]: Updated to be relevant, lmk if it isn't descriptive enough

File: views/ABViewCore.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (Achoobert) [reply views/ABViewCore.js:801]: added comment

File: labels/en.js [hunk 7]
```diff
@@ -1,201 +1,201 @@
 /* eslint-disable */
 module.exports = {
    /* key : label */
-   "Year" : "Year",
```
Comment (Hiro-Nakamura) [inline labels/en.js:4]: I think this is one of those `eslint` / `prettier` config issues.  can you verify your IDE is using our settings for this file?

Comment (Hiro-Nakamura) [reply labels/en.js:4]: I think there `should` be the " " in there.

**Review by Hiro-Nakamura — CHANGES_REQUESTED** [review]
Yeah, let's clean up those Labels and then this is good to go.

File: labels/en.js (hunk 7 above)
Comment (Achoobert) [reply labels/en.js:4]: Fixed

**Pull request merged**